src/shared/lib/utils/survey.ts

```diff
@@ -13,4 +15,208 @@ export const invertColor = (hex: string) => {
+  inputType: 'ageSelector';
+  answer: number;
+};
+
+type SliderInput = {
+  inputType: 'slider';
+  answer: number;
+};
+
+type CheckBoxInput = {
+  inputType: 'checkbox';
+  answer: string[];
+};
+
+type Answer = DateAnswer | TimeRangeAnswer | string | string[] | number;
+
+type InputAnswer =
+  | DateInput
+  | TimeRangeInput
+  | TextInput
+  | RadioInput
+  | CheckBoxInput
+  | AgeSelectorInput
+  | SliderInput;
+
+type ActivityItems = {
+  variableName: string;
+  inputType: InputAnswer['inputType'];
+  valueConstraints: {
+    itemList: { name: string; value: string | number }[];
+  };
+}[];
+
+export class MarkdownVariableReplacer {
+  private markdown = '';
+  private readonly activityItems: ActivityItems;
+  private readonly answers;
+
+  constructor(
+    markdown: string,
+    activityItems: ActivityItems,
+    answers: Answer[],
+  ) {
+    this.markdown = markdown;
+    this.activityItems = activityItems;
+    this.answers = answers;
+  }
+
+  private extractVariables = (markdown: string): string[] => {
+    const regEx = /\[\[(.*?)]]/g;
+    const matches = [];
+    let found;
+    while ((found = regEx.exec(markdown))) {
+      matches.push(found[1]);
+    }
+    return matches;
+  };
+
+  private updateMarkdown = (variableName: string, replaceValue: string) => {
+    const reg = new RegExp(`\\[\\[${variableName}\\]\\]`, 'gi');
+    this.markdown = this.markdown.replace(reg, replaceValue);
+  };
+
+  private formatTime = (
+    timeObject: TimeRangeAnswer['to'] | undefined,
+  ): string => {
+    if (!timeObject) {
+      return '';
+    }
+    const { hour, minute } = timeObject;
+    return format(new Date(0, 0, 0, hour, minute), 'HH:mm');
+  };
+
+  private formatDate = (dateObject: undefined | DateAnswer): string => {
+    if (!dateObject) {
+      return '';
+    }
+    const { year, month, day } = dateObject;
+    return format(new Date(year, month, day), 'y-MM-dd');
+  };
+
+  public process = () => {
+    const variableNames = this.extractVariables(this.markdown);
+
+    if (!this.answers?.length || !variableNames?.length) {
+      return this.markdown;
+    }
+
+    try {
+      variableNames.forEach(variableName => {
+        const updated = this.getReplaceValue(variableName);
+        this.updateMarkdown(variableName, updated);
+      });
+    } catch (error) {
+      console.warn(error);
+    }
+
+    return this.markdown;
+  };
+
+  private escapeSpecialChars = (value: number | string) => {
+    return value.toString().replace(/(?=[$&])/g, '\\');
+  };
+
+  private getReplaceValue = (variableName: string) => {
+    const foundIndex = this.activityItems.findIndex(
+      item => item.variableName === variableName,
+    );
+    const answerNotFound = foundIndex < 0 || !this.answers[foundIndex];
+
+    if (answerNotFound) {
+      return `[[${variableName}]]`;
+    }
+
+    const activityItem = this.activityItems[foundIndex];
+    let updated = '';
+    const inputAnswer: InputAnswer = <InputAnswer>{
```

Comment:
Is it necessary to put <InputAnswer> generic here?

Reply:
@BamMironov yes, it gives

TS2322: Type '{ inputType: "timeRange" | "date" | "text" | "radio" | "ageSelector" | "slider" | "checkbox"; answer: Answer; }' is not assignable to type 'InputAnswer'.   Type '{ inputType: "timeRange" | "date" | "text" | "radio" | "ageSelector" | "slider" | "checkbox"; answer: Answer; }' is not assignable to type 'CheckBoxInput'.     Types of property 'inputType' are incompatible.       Type '"timeRange" | "date" | "text" | "radio" | "ageSelector" | "slider" | "checkbox"' is not assignable to type '"checkbox"'.         Type '"timeRange"' is not assignable to type '"checkbox"